types/stylelint/index.d.ts

```diff
@@ -246,7 +246,7 @@ declare namespace stylelint {
 	/** @internal */
 	export type RuleBase<P = any, S = any> = (
 		primaryOption: P,
-		secondaryOptions: Record<string, S>,
+		secondaryOptions: S extends Record<string, any> ? S : never,
```

Comment:
```suggestion
		secondaryOptions: S,
```

see https://github.com/stylelint/stylelint/pull/7950#discussion_r1730886345